review-problems.ml

```diff
@@ -917,3 +917,203 @@ module Lazy = struct
+  The function `apply_to_list` should take a function `f` that maps values of type `'a to 'b`, and a list of values of type `'a`. 
+  It should return a list of values of type `'b`.
+
+- Write at least two example functions `f1` and `f2`, each with different type signatures. For instance, `f1` could be a function 
+  that squares an integer, and `f2` could be a function that converts a string to uppercase.
+
+- Apply the apply_to_list function to each of the example functions f1 and f2, and a list of values. 
+  Provide examples of the input list and the expected output for each function.
+
+*)
+
+
+(*
+QUESTION#4 :-
+
+Imagine you have a list of students, each represented by their name, a transcript of courses and grades
+and you want to find students who have achieved high honors. 
+
+Using higher-order functions in OCaml, solve the following problems:
+
+  - Implement a function `find_honors_student` with the following signature:
+      
+      val find_honors_student : student list -> name option
+
+      This function should take a list of students and return the name of a student who has 
+      achieved a grade of over 90 in all their courses. If there is no such student, return `None`.
+
+  - Implement a function `average_grade` with the following signature:
+
+      val average_grade : student -> float
+
+      This function should take a single student and calculate their average grade across all courses. 
+      Use this function to find the student with the highest average grade in the list of students.
+
+  - Implement a function `top_students` with the following signature:
+
+      val top_students : student list -> name list
+
+      This function should take a list of students and return a list of names of students who have achieved a 
+      grade of 95 or higher in at least one course.
+
+*)
+
+(*
+QUESTION#5 :-
+
+Recursive Chessboard Exploration
+
+Consider an `n` x `n` chessboard represented as a grid of squares. Each square can be either empty or blocked. 
+You start at the top-left square `(0, 0)` and want to reach the bottom-right square `(n-1, n-1)`. 
+However, there are some restrictions:
+
+- You can only move right or down.
+- You cannot move through blocked squares.
+- You cannot visit the same square twice.
+
+(1). Implement a recursive function `explore_chessboard` with the following signature:
+
+  val explore_chessboard : int -> bool array array -> bool
+
+      This function should take an integer `n` (the size of the chessboard), a 2D boolean array representing the 
+      blocked squares (where `true` indicates a blocked square), and return `true` if there is a path from the top-left 
+      corner `(0, 0)` to the bottom-right corner `(n-1, n-1)` that adheres to the rules mentioned above. Otherwise, return `false`.
+
+(2). Create a higher-order function `process_path` with the following signature:
+
+  val process_path : (int * int -> unit) -> int -> bool array array -> unit
```

Comment:
Midterm 1 doesn't cover any effectful programming, so students would not in general be familiar with functions returning `unit` that are expected to perform effects.

The name and description of the function are a bit misleading too: `process_path` sounds like it should accept a _path_ as input, but from looking at the solution, the given function is called by `process_path` on the coordinates of many squares, not arranged into a path.

Also, the `int` parameter is redundant with the information in the array, no? In other words, the size of the board doesn't need to be separately specified because it can be calculated from the board.

I'm not sure how to adjust this problem to be helpful study for midterm 1.